etcd-defrag/etcddefrag_controller.go

```diff
@@ -0,0 +1,310 @@
+package etcddefrag
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"os"
+	"time"
+
+	"github.com/go-logr/logr"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	"github.com/openshift/hypershift/pkg/etcdcli"
+	"github.com/openshift/hypershift/support/upsert"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	pollWaitDuration                       = 2 * time.Second
+	pollTimeoutDuration                    = 60 * time.Second
+	maxDefragFailuresBeforeDegrade         = 3
+	minDefragBytes                 int64   = 100 * 1024 * 1024 // 100MB
+	minDefragWaitDuration                  = 36 * time.Second
+	maxFragmentedPercentage        float64 = 45
+
+	controllerRequeueDuration = 10 * time.Minute
+)
+
+type DefragController struct {
+	client.Client
+	log logr.Logger
+
+	ControllerName string
+	HCPNamespace   string
+	upsert.CreateOrUpdateProvider
+
+	etcdClient         etcdcli.EtcdClient
+	numDefragFailures  int
+	requeueTime        time.Time
+	defragWaitDuration time.Duration
+}
+
+func (r *DefragController) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+
+	b := ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedControlPlane{}).
+		WithOptions(controller.Options{
+			RateLimiter: workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 10*time.Second),
+		}).Named(r.ControllerName)
+
+	r.log.Info("NewControllerManagedBy() called")
+	if _, err := b.Build(r); err != nil {
+		return err
+	}
+
+	endpointsFunc := func() ([]string, error) {
+		return r.etcdEndpoints(ctx)
+	}
+	r.etcdClient = etcdcli.NewEtcdClient(endpointsFunc, events.NewLoggingEventRecorder(r.ControllerName))
+
+	// Set this so that it will immediately requeue itself.
+	r.requeueTime = time.Now().Add(controllerRequeueDuration)
+	r.defragWaitDuration = minDefragWaitDuration
+
+	r.log.Info("Running mgr.Start()")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
```

Comment:
`mgr.Start` is usually called from main.
This function shouldn't be doing `os.Exist()`, but return an error instead.

Reply:
Ah yes, I see.